internal/messagequeue/messagequeue.go

```diff
@@ -327,6 +371,62 @@ func (mq *MessageQueue) sendMessage() {
 	}
 }
 
+// If the peer is running an older version of Bitswap that doesn't support the
+// DONT_HAVE response, watch for timeouts on any want-blocks we sent the peer,
+// and if there is a timeout simulate a DONT_HAVE response.
+func (mq *MessageQueue) simulateDontHaveWithTimeout(msg bsmsg.BitSwapMessage) {
+	// If the peer supports DONT_HAVE responses, we don't need to simulate
+	if mq.sender.SupportsHave() {
```

Comment:
Do we want to assume that peers will behave themselves? IMO, we should probably do this for _all_ peers. "haves" and "don't haves" just let us react _faster_ without waiting a timeout.

Note: we'll need to tune the timeout, ideally based on the latency to the peer... we might want to use the libp2p ping protocol to determine that?

Reply:
My thinking is that
- if we get a timeout from a peer running an older version Bitswap we can treat it as a DONT_HAVE
- if we get a timeout from a peer running new Bitswap we should assume that the peer is overloaded / there are network issues and we should remove the peer from the session (not yet implemented)

Does that make sense to you?

Reply:
Also I don't know much about the libp2p ping protocol, can you describe how we would use it to determine latency? If we're already keeping track of timing information to simulate DONT_HAVEs we may also be able to use that to work out a rough measure of latency.

Reply:
> Does that make sense to you?

Yes. Sorry, I completely forgot that we discussed this. Recalling that discussion, the conclusion was that we should kick peers from the session when:

1. They've been idle for some period of time.
2. They have outstanding wants with "don't have" flags.

Reply:
>  If we're already keeping track of timing information to simulate DONT_HAVEs we may also be able to use that to work out a rough measure of latency.

https://github.com/libp2p/go-libp2p/blob/master/p2p/protocol/ping/ping.go#L86

Really, that might be overkill. I only suggested it because we need _some_ way to learn latency and I'd like to simplify tracking it as much as possible. However, really, we want to know how long it's taking our peer's _bitswap_ service to respond to requests, so using the ping service isn't as accurate as just measuring how long it takes our peer to respond to wants.

Reply:
Thinking a little more about this - it might actually be useful to use the ping service because it allows us to know roughly how long to wait for a response the first time we send a message to the peer.

If it's a peer running the old Bitswap, and the ping time is 5ms then the first time we send a want-block to the peer we can wait 100ms instead of 5s before assuming DONT_HAVE.

Reply:
Discussed out of band: continuously pinging might be too expensive. Maybe just ask libp2p what the latency is and, if we don't know, try pinging (once).

Reply:
In [014348b](https://github.com/ipfs/go-bitswap/pull/248/commits/014348b08e989b05435d200736fe453e9567eaed) I added a pinger that will
- attempt to ping the peer
- wait to set timers until the peer has responded to the ping
- ping periodically thereafter (by default every 30s) and keep a running average

Do you think the periodic ping is overkill?

Benchmark after making this change:
```
go test . -run a -bench BenchmarkFetchFromOldBitswap
goos: darwin
goarch: amd64
pkg: github.com/ipfs/go-bitswap
BenchmarkFetchFromOldBitswap/3Nodes-Overlap3-OneAtATime-8         	      25	  44776526 ns/op
BenchmarkFetchFromOldBitswap/3Nodes-AllToAll-OneAtATime-8         	1000000000	         0.234 ns/op
BenchmarkFetchFromOldBitswap/3Nodes-Overlap3-AllConcurrent-8      	1000000000	         0.0269 ns/op
BenchmarkFetchFromOldBitswap/3Nodes-Overlap3-OneAtATime (5 runs / 3.37s):   0.674s: sent 22, recv 11, dups 1 / 11
BenchmarkFetchFromOldBitswap/3Nodes-AllToAll-OneAtATime (14 runs / 3.21s):  0.230s: sent 20, recv 11, dups 1 / 11
BenchmarkFetchFromOldBitswap/3Nodes-Overlap3-AllConcurrent (7 runs / 0.17s): 0.025s: sent 2, recv 6, dups 4 / 14
PASS
ok  	github.com/ipfs/go-bitswap	8.137s
```

Reply:
IMO, the periodic ping is probably overkill. I'd:

1. Check to see if we have a known latency.
2. If not, ping once.

That's probably enough for now.

If we decide to not track want latency, and continue with pings, then we can get fancy.